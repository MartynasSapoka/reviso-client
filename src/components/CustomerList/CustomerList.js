import React from 'react';
import PropTypes from 'prop-types';
import JobsTableHeader from "./components/CustomerTableHeader/CustomerTableHeader";
import CustomerRow from "./components/CustomerRow/CustomerRow";
import SearchFieldContainer from "../SearchField/SearchFieldContainer";
import InsertNewRow from "./components/InsertNewRow";

class CustomerList extends React.Component {

  constructor(props) {
    super(props);
    this.state = {insertNew: false}
  }

  render() {
    const {customers} = this.props;
    return(
      <div className="container">
        <div className="col-xs-12">
          <SearchFieldContainer/>
          <table className="table table-striped jobs-table">
            <JobsTableHeader colNames={['Company', 'Total Hours', 'Total Amount']}/>
            <tbody className="text-left">
            {customers.map(j =>
              <CustomerRow customer={j} key={j._id} removeCustomer={this.props.removeCustomer}/>
            )}
            {this.state.insertNew && <InsertNewRow onSave={this.onCreateCustomer}/>}
            </tbody>
          </table>
          <button className="btn btn-default" onClick={() => this.setState({insertNew: true})}>Add</button>
        </div>
      </div>
    )
  }

  onCreateCustomer = (customer) => {
    this.props.createCustomer(customer);
    this.setState({insertNew: false})
  };
}

CustomerList.propTypes = {
  customers: PropTypes.array.isRequired,
  createCustomer: PropTypes.func.isRequired,
  removeCustomer: PropTypes.func.isRequired
};

export default CustomerList;