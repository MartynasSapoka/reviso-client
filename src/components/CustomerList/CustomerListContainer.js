import React from "react";
import {connect} from 'react-redux'
import ClientsTable from './CustomerList';
import {createCustomer, fetchAllCustomers, removeCustomer} from "../../actions/CustomerActions";

class CustomerListContainer extends React.Component {
  componentDidMount() {
    this.props.fetchData();
  }

  render() {
    const {fetchData, ...rest} = this.props;
    return <ClientsTable {...rest}/>
  }
}


const mapStateToProps = state => {
  return {
    customers: state.filter.text ? state.customers.filter(c => c.name.toLowerCase().includes(state.filter.text.toLowerCase())) : state.customers
  }
};

CustomerListContainer = connect(
  mapStateToProps,
  dispatch => ({
    fetchData: () => dispatch(fetchAllCustomers),
    createCustomer: customer => dispatch(createCustomer(customer)),
    removeCustomer: (customer) => dispatch(removeCustomer(customer))
  })
)(CustomerListContainer);

export default CustomerListContainer;