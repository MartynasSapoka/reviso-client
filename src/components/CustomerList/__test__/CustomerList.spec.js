import React from 'react';
import {shallow, mount, render} from 'enzyme';
import CustomerList from "../CustomerList";
import toJson from 'enzyme-to-json';

const customerList = [
  {
    "billings":
      [{
        "timestamp": "2018-03-18T19:19:00.496Z",
        "hours": 12,
        "hourlyRate": 8,
        "_id": "5aaebad500458185fb51e4cb"
      }],
    "_id": "5aaea959dd0322810727413c",
    "name": 'Test'
  },
  {
    "billings": [
      {
        "timestamp": "2018-03-12T00:00:00.000Z",
        "hours": 12,
        "hourlyRate": 10,
        "_id": "5ab0147d84a389b5aa597032"
      },
      {
        "timestamp": "2018-03-07T00:00:00.000Z",
        "hours": 42,
        "hourlyRate": 9,
        "_id": "5ab014b384a389b5aa597033"
      }],
    "_id": "5aaea9f2dd0322810727413e",
    "name": "Reviso",
  }
];

describe('CustomerList', () => {
  const removeFnMock = jest.fn();
  const createFnMock = jest.fn();
  it('Renders without crashing', () => {
    const wrapper = shallow(<CustomerList removeCustomer={removeFnMock} createCustomer={createFnMock}
                                          customers={customerList}/>)
  });

  it('Matches snapshot', () => {
    const wrapper = shallow(<CustomerList removeCustomer={removeFnMock} createCustomer={createFnMock} customers={customerList}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  })
});