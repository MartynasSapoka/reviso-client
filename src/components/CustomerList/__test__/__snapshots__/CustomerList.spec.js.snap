// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerList Matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="col-xs-12"
  >
    <Connect(SearchField) />
    <table
      className="table table-striped jobs-table"
    >
      <CustomerTableHeader
        colNames={
          Array [
            "Company",
            "Total Hours",
            "Total Amount",
          ]
        }
      />
      <tbody
        className="text-left"
      >
        <CustomerRow
          customer={
            Object {
              "_id": "5aaea959dd0322810727413c",
              "billings": Array [
                Object {
                  "_id": "5aaebad500458185fb51e4cb",
                  "hourlyRate": 8,
                  "hours": 12,
                  "timestamp": "2018-03-18T19:19:00.496Z",
                },
              ],
              "name": "Test",
            }
          }
          key="5aaea959dd0322810727413c"
          removeCustomer={[Function]}
        />
        <CustomerRow
          customer={
            Object {
              "_id": "5aaea9f2dd0322810727413e",
              "billings": Array [
                Object {
                  "_id": "5ab0147d84a389b5aa597032",
                  "hourlyRate": 10,
                  "hours": 12,
                  "timestamp": "2018-03-12T00:00:00.000Z",
                },
                Object {
                  "_id": "5ab014b384a389b5aa597033",
                  "hourlyRate": 9,
                  "hours": 42,
                  "timestamp": "2018-03-07T00:00:00.000Z",
                },
              ],
              "name": "Reviso",
            }
          }
          key="5aaea9f2dd0322810727413e"
          removeCustomer={[Function]}
        />
      </tbody>
    </table>
    <button
      className="btn btn-default"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
</div>
`;
